python/pyspark/worker.py

```diff
@@ -133,65 +134,103 @@ def verify_result_length(result, length):
     )
 
 
-def wrap_batch_iter_udf(f, return_type):
+def wrap_batch_iter_udf(f, return_type, is_arrow_iter=False):
     arrow_return_type = to_arrow_type(return_type)
+    iter_type_label = (
+        "pyarrow.RecordBatch"
+        if is_arrow_iter
+        else ("pandas.DataFrame" if type(return_type) == StructType else "pandas.Series")
+    )
 
-    def verify_result_type(result):
-        if not hasattr(result, "__len__"):
-            pd_type = "Pandas.DataFrame" if type(return_type) == StructType else "Pandas.Series"
+    def verify_result(result):
+        if not isinstance(result, Iterator) and not hasattr(result, "__iter__"):
             raise TypeError(
                 "Return type of the user-defined function should be "
-                "{}, but is {}".format(pd_type, type(result))
+                "iterator of {}, but is {}".format(iter_type_label, type(result))
             )
         return result
 
+    def verify_element(elem):
+        if is_arrow_iter:
+            import pyarrow as pa
+
+            if not isinstance(elem, pa.RecordBatch):
+                raise TypeError(
+                    "Return type of the user-defined function should be "
+                    "iterator of {}, but is iterator of {}".format(iter_type_label, type(elem))
+                )
+        else:
+            import pandas as pd
+
+            if not isinstance(elem, pd.DataFrame if type(return_type) == StructType else pd.Series):
+                raise TypeError(
```

Comment:
ditto, and other places as well ?

Basically we should use the PySpark specific errors instead of Python build-in exceptions.

See https://github.com/apache/spark/blob/master/python/pyspark/errors/__init__.py for more details about PySpark specific errors.

Reply:
Sure, but the whole file uses `TypedError` instead of the new `PySparkTypeError`.

I'd rather fix that in a separate PR and not fixing unrelated code in this PR.

Reply:
Of course. You can only address the related changes

Reply:
Done